flink-libraries/flink-table/src/test/scala/org/apache/flink/table/formats/TableFormatFactoryServiceTest.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.formats
+
+import java.util.{HashMap => JHashMap, Map => JMap}
+
+import org.apache.flink.table.api.{AmbiguousTableFormatException, NoMatchingTableFormatException}
+import org.apache.flink.table.formats.utils.TestAmbiguousTableFormatFactory
```

Comment:
Am I guessing correctly, that this test is searching the class path for classes that extend `TableFormatFactory` and tries to instantiate/match them to the properties below? This reminds me too much of testing static code or code with global variables. Also it probably shows problem with encapsulation/code reusability. Shouldn't `TableFormatFactoryService` be split into following components:
1. `TableFormatFactoryDiscoverer` or even more generic `new ClassDiscoverer(classOf[TableFormatFactory])` - class that searches for the instances of `TableFormatFactory` and returns them as a list
2. `new TableFormatFactoryService(formats: Seq[Class[TableFormatFactory]])` that validates properties and instantiate format factories
3. some tiny class/code that glues 1. and 2. together

testing those components would be easier and less implicit compared to how it is now. 

For example now, it is unclear to me which format factories affect this test and in what way.

Reply:
Yes, we could split the logic here. Would it be ok for you do do this in a follow-up issue? I guess we can combine the logic with `TableSourceFactory` and the upcoming sink discovery. We could introduce a general `FactoryDiscoverer` for all interfaces.